### PR TITLE
Data store

### PR DESCRIPTION
**Pull request by nathanidp**

### CODE REVIEW

File: modules/data-store/src/entity.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const SerializableClass = require("abitbol-serializable");
+
+/**
+ * A structure that can be added to a ProjectManager.
+ *
+ * @class obsidian-project.lib.structure
+ * @extends abitbol-serializable
+ */
+const Structure = SerializableClass.$extend({
+    __name__: "Structure",
```
Comment (flozz) [inline modules/data-store/src/entity.js]: Maybe you should call id `entity`?

File: modules/data-store/src/entity.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * @property store
+     * @readOnly
+     */
+    getStore() {
+        return this.$data._store; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/entity.js]: I think we should use a `Symbol` here to be sure to not collide with an user-defined property in a sub-class

File: modules/data-store/src/entity.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    },
+
+    /**
+     * @property layer
+     * @readOnly
+     * @type obsidian-project.lib.structure[]
+     */
+    getPath() {
+        return this.$data._path; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/entity.js]: Same, a `Symbol` would be better

File: modules/data-store/src/data-store.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+
+    /**
+     * Remove an entity to the store.
+     *
+     * @param {SerializableClass|string} entity The entity or its id
+     * @return {undefined}
+     */
+    removeEntity(entity) {
+        // const id = (entity instanceof String) ? entity : entity.id;
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: You can remove this dead code ;)

File: modules/data-store/src/data-store.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            realEntity = entity;
+        }
+        const path = this.entitiesByUuid[id].getPath();
+        delete this.entitiesByUuid[id];
+        const index = this.entitiesByPath[path].indexOf(realEntity);
+        this.entitiesByPath[path].splice(index, 1);
+        if (this.entitiesByPath[path].length === 0) {
+            delete this.entitiesByPath[path];
+        }
+        delete realEntity.$data._path; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: I think you should not delete the property but set it to `null` (also, this attribute should probably be initialized to `null` in the Entity constructor)

File: modules/data-store/src/data-store.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+     */
+    removeEntity(entity) {
+        // const id = (entity instanceof String) ? entity : entity.id;
+        let id;
+        let realEntity;
+        if ((typeof entity) === "string") {
+            id = entity;
+            realEntity = this.entitiesByUuid[id];
+        } else {
+            id = entity.id; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: We may have to remove the eslint rule that seems too boring... Also prefer disabling only ONE rule on the line instead of all rules (`eslint-disable-line prefer-destructuring`)

File: modules/data-store/src/data-store.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        entity.$data._path = path; // eslint-disable-line
+        this.entitiesByPath[path].push(entity);
+        this.entitiesByUuid[entity.id] = entity;
+        // TODO add "path" property to entity and assign the path value
+        // TODO emit an event ("entity-added" with the entity in param)
+    }
+
+
+    /**
+     * Remove an entity to the store.
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: → Remove an entity **from** the store
(yes that's my mistake... #copyPaste) :)

File: modules/data-store/src/data-store.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+    constructor() {
+        this.entitiesByPath = {};
+        this.entitiesByUuid = {};
+    }
+
+    /**
+     * Add an entity to the store.
+     *
+     * @param {SerializableClass} entity The entity
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: As you created a class called `Entity`, you should update the doc to replace `SerializableClass` by `Entity` :)

(There is other places in the file where this should be changed)

File: modules/data-store/src/data-store.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * List entities that matches the given path. (e.g.: ``"/"``, ``"/*"``
+     * ``"/models/*"``).
+     *
+     * @param {string} path The path to list (accepts globing).
+     * @return {array[string]} The ID of the mathing entities.
+     */
+    listEntities(path = "/**") {
+        // TODO
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: what remains TODO here? :)

File: modules/data-store/src/data-store.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+     *
+     * @return {undefined}
+     */
+    clear() {
+        const keys = Object.keys(this.entitiesByPath);
+        keys.forEach(
+            (key) => {
+                this.entitiesByPath[key].forEach(
+                    (entity) => {
+                        delete this.entitiesByUuid[entity.id];
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: Maybe it is better to call the `removeEntity()` method here, this will assure a consistent event emission and objects cleaning

File: modules/data-store/src/data-store.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    /**
+     * Serialize all stored entities.
+     *
+     * @return {string} The entities serialized as JSON.
+     */
+    serializeEntities() {  // → string (JSON)
+        const serialized = JSON.stringify(serializer.objectSerializer(this.entitiesByPath));
+        return serialized;
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: If there is only one line, there is no need of an intermediate variable

File: modules/data-store/src/data-store.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    /**
+     * Unserialize entities from given JSON (the unserialized entities are
+     * added to the store).
+     *
+     * @param {string} serializedEntities The entities serialized as JSON
+     * @return {undefined}
+     */
+    unserializeEntities(serializedEntities) {
+        const unserialized = serializer.objectUnserializer(JSON.parse(serializedEntities));
+        Object.keys(unserialized).forEach((key) => {
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: maybe you can rename `key` to `path` here?

File: modules/data-store/test/data-store.test.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        const object = {
+            "/tata": [entity],
+        };
+        const object2 = {
+            "/toto": [entity],
+        };
+        expect(dataStore.entitiesByPath).toEqual(object);
+        expect(dataStore.entitiesByPath).not.toEqual(object2);
+    });
+    test("many entities in same path", () => {
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: Can you add blank line between tests ? :)

File: modules/data-store/src/data-store.js [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+
+/**
+ * Stores project's entities and blobs.
+ */
+class DataStore {
+
+    constructor() {
+        this.entitiesByPath = {};
+        this.entitiesByUuid = {};
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: I think you should use `Symbol`s here to avoid access to those properties as they should not be modified from outside of this class

Comment (nathanidp) [reply modules/data-store/src/data-store.js]: It was for the tests, should I put a getter ?

File: modules/data-store/test/data-store.test.js (hunk 13 above)
Comment (nathanidp) [reply modules/data-store/test/data-store.test.js]: Done

File: modules/data-store/src/data-store.js (hunk 14 above)
Comment (flozz) [reply modules/data-store/src/data-store.js]: You can export symbols and use them in tests :)

File: modules/data-store/src/data-store.js (hunk 12 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: Done

File: modules/data-store/src/entity.js (hunk 1 above)
Comment (nathanidp) [reply modules/data-store/src/entity.js]: Done

File: modules/data-store/src/data-store.js (hunk 4 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: Done

File: modules/data-store/src/data-store.js (hunk 6 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: Done

File: modules/data-store/src/data-store.js (hunk 7 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: Fixed

File: modules/data-store/src/data-store.js (hunk 8 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: Done

File: modules/data-store/src/data-store.js (hunk 9 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: DO

Comment (nathanidp) [reply modules/data-store/src/data-store.js]: NE

File: modules/data-store/src/data-store.js (hunk 10 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: OK

File: modules/data-store/src/data-store.js (hunk 11 above)
Comment (nathanidp) [reply modules/data-store/src/data-store.js]: DONE

File: modules/data-store/src/entity.js [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+     * @property path
+     * @readOnly
+     * @type string
+     */
+    getPath() {
+        return this.$data[ENTITY_PATH];
+    },
+
+    /**
+     * Destroys the entity from the project
```
Comment (flozz) [inline modules/data-store/src/entity.js]: > Destroys the entity and removes it from the project 

:)

File: modules/data-store/src/entity.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    __init__(params) {
+        this.$data[DATA_STORE] = null;
+        this.$data[ENTITY_PATH] = null;
+        this.$super(params);
+    },
+
+    /**
+     * @property store
+     * @readOnly
+     */
```
Comment (flozz) [inline modules/data-store/src/entity.js]: Please add a description and a `@type`

File: modules/data-store/src/entity.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    getStore() {
+        return this.$data[DATA_STORE];
+    },
+
+    /**
+     * @property path
+     * @readOnly
+     * @type string
+     */
+    getPath() {
```
Comment (flozz) [inline modules/data-store/src/entity.js]: Please add a description here too :)

File: modules/data-store/src/data-store.js [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+     * @param {Entity} entity The entity
+     * @param {string} path Path where the entity will be stored
+     *                      (default=``"/"``)
+     * @return {undefined}
+     */
+    addEntity(entity, path = "/") {
+        if (!this[ENTITIES_BY_PATH][path]) {
+            this[ENTITIES_BY_PATH][path] = [];
+        }
+        entity.$data[ENTITY_PATH] = path; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: → `// eslint-disable-line no-param-reassign`

File: modules/data-store/src/data-store.js [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+            realEntity = entity;
+        }
+        const path = this[ENTITIES_BY_UUID][id].getPath();
+        delete this[ENTITIES_BY_UUID][id];
+        const index = this[ENTITIES_BY_PATH][path].indexOf(realEntity);
+        this[ENTITIES_BY_PATH][path].splice(index, 1);
+        if (this[ENTITIES_BY_PATH][path].length === 0) {
+            delete this[ENTITIES_BY_PATH][path];
+        }
+        delete realEntity.$data[ENTITY_PATH]; // eslint-disable-line
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: * no need to disable eslint here.
* Do not delete the property, just set it to `null`.
* You should also set the store reference to `null` too.

File: modules/data-store/src/data-store.js [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+        let id;
+        let realEntity;
+        if ((typeof entity) === "string") {
+            id = entity;
+            realEntity = this[ENTITIES_BY_UUID][id];
+        } else {
+            id = entity.id; // eslint-disable-line prefer-destructuring
+            realEntity = entity;
+        }
+        const path = this[ENTITIES_BY_UUID][id].getPath();
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: As it is an Abitbol class, you can just use `entity.path` instead of `entity.getPath()` ;)

File: modules/data-store/src/data-store.js [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+        }
+        delete realEntity.$data[ENTITY_PATH]; // eslint-disable-line
+        // TODO emit an event ("entity-removed" with the entity in param)
+    }
+
+    /**
+     * Get an entity
+     *
+     * @param {string} id The entity ID.
+     */
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: Please add a `@return {Entity|undefined}` section

File: modules/data-store/src/data-store.js [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+    getEntity(id) {
+        return this[ENTITIES_BY_UUID][id];
+    }
+
+    /**
+     * List entities that matches the given path. (e.g.: ``"/"``, ``"/*"``
+     * ``"/models/*"``).
+     *
+     * @param {string} path The path to list (accepts globing).
+     * @return {array[string]} The ID of the mathing entities.
```
Comment (flozz) [inline modules/data-store/src/data-store.js]: The return type does not match the implementation, fix the doc :)

also there is a typo: mathing → matching

File: modules/data-store/test/data-store.test.js [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+        const entity = new Entity({ id: "entity-0" });
+        dataStore.addEntity(entity, "/tata");
+        const object = {
+            "/tata": [entity],
+        };
+        const object2 = {
+            "/toto": [entity],
+        };
+        expect(dataStore[ENTITIES_BY_PATH]).toEqual(object);
+        expect(dataStore[ENTITIES_BY_PATH]).not.toEqual(object2);
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: I think this test is useless

File: modules/data-store/test/data-store.test.js [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+        expect(dataStore[ENTITIES_BY_PATH]["/tata"][index]).toBe(entity);
+    });
+
+    test("entity path is correct", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        dataStore.addEntity(entity, "/tata");
+        expect(entity.getPath()).toBe("/tata");
+    });
+
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: meybe add a tests that check that the store reference is set too

File: modules/data-store/test/data-store.test.js [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+describe("DataStore.removeEntity", () => {
+
+    test("entity is removed from store", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        const entity1 = new Entity({ id: "entity-1" });
+        dataStore.addEntity(entity, "/tata");
+        dataStore.addEntity(entity1, "/tata");
+        dataStore.removeEntity(entity.id);
+        expect(dataStore[ENTITIES_BY_PATH]["/tata"].indexOf(entity)).toEqual(-1);
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: use `.not.toContain()` → https://jestjs.io/docs/en/expect#tocontainitem ;)

File: modules/data-store/test/data-store.test.js [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+    });
+
+    test("entity path is removed", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        dataStore.addEntity(entity, "/tata");
+        dataStore.removeEntity(entity.id);
+        expect(entity.getPath()).toBeUndefined();
+    });
+
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: also test that the entity do not reference the store anymore

File: modules/data-store/test/data-store.test.js [hunk 27]
```diff
@@ -0,0 +1,239 @@
...
+describe("DataStore.getEntity", () => {
+
+    test("entity exists", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        dataStore.addEntity(entity);
+        expect(dataStore.getEntity(entity.id)).toBe(entity);
+    });
+
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: also test by giving the entity ID

File: modules/data-store/test/data-store.test.js [hunk 28]
```diff
@@ -0,0 +1,239 @@
...
+    test("path is correct", () => {
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: This test is not related to getEntity(), it should (and it is already) be tested in the addEntity() section

File: modules/data-store/test/data-store.test.js [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+        dataStore.addEntity(entity, "/tata");
+        dataStore.addEntity(entity1, "/tata/toto");
+        const serialized = dataStore.serializeEntities();
+        expect(typeof serialized).toBe("string");
+        expect(typeof JSON.parse(serialized)).toBe("object");
+    });
+
+});
+
+describe("DataStore.UnserializeEntities", () => {
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: U to lower case ;)

File: modules/data-store/test/data-store.test.js [hunk 30]
```diff
@@ -0,0 +1,239 @@
...
+            "/toto",
+        ];
+        const serializedData = dataStoreSerializer.serializeEntities();
+        dataStoreUnserializer.unserializeEntities(serializedData);
+        const keys = Object.keys(dataStoreUnserializer[ENTITIES_BY_PATH]);
+        expect(keys).toEqual(expect.arrayContaining(object));
+        expect(keys.length).toEqual(object.length);
+        expect(dataStoreUnserializer[ENTITIES_BY_PATH]["/tata"].length).toEqual(2);
+        expect(dataStoreUnserializer[ENTITIES_BY_PATH]["/tata/titi"].length).toEqual(1);
+        expect(dataStoreUnserializer[ENTITIES_BY_PATH]["/toto"].length).toEqual(1);
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: use `.toHaveLength()` instead → https://jestjs.io/docs/en/expect#tohavelengthnumber

File: modules/data-store/test/data-store.test.js [hunk 31]
```diff
@@ -0,0 +1,239 @@
...
+    });
+
+    test("can remove entity by giving entity in parameters", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        const entity1 = new Entity({ id: "entity-1" });
+        dataStore.addEntity(entity, "/tata");
+        dataStore.addEntity(entity1, "/tata");
+        dataStore.removeEntity(entity);
+        expect(dataStore[ENTITIES_BY_PATH]["/tata"].indexOf(entity)).toEqual(-1);
```
Comment (flozz) [inline modules/data-store/test/data-store.test.js]: /!\ → `.not.toContain()`

File: modules/data-store/test/data-store.test.js [hunk 32]
```diff
@@ -0,0 +1,239 @@
...
+
+});
+
+describe("DataStore.getEntity", () => {
+
+    test("entity exists", () => {
+        const dataStore = new DataStore();
+        const entity = new Entity({ id: "entity-0" });
+        dataStore.addEntity(entity);
+        expect(dataStore.getEntity(entity.id)).toBe(entity);
```
Comment (nathanidp) [inline modules/data-store/test/data-store.test.js]: Check if undefined
